File: frontend/src/api/social.ts
import { apiClient } from "./client";
import type { ExecutionJob, ExecutionJobCreateResult } from "../types/execution";
import type {
  PublishArtifactCandidate,
  PublishTask,
  SocialAccount,
  InstagramConnectResult,
  TikTokConnectResult,
  InstagramFinalizePreflight,
  InstagramPublishPreflight,
  InstagramPublishingMetadata,
  YouTubeConnectResult,
  YouTubePreflight,
  YouTubePublishingMetadata,
  TikTokCreatorInfoSnapshot,
  TikTokPublishingMetadata,
  TikTokPublishPreflight,
} from "../types/social";

export async function connectYouTube(
  productId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<YouTubeConnectResult>(
    "/social-accounts/youtube/connect",
    { product_id: productId },
    { signal },
  );
  return response.data;
}

export async function connectInstagram(productId: number, signal?: AbortSignal) {
  const response = await apiClient.post<InstagramConnectResult>(
    "/social-accounts/instagram/connect",
    { product_id: productId },
    { signal },
  );
  return response.data;
}

export async function disconnectInstagramAccount(
  productId: number,
  accountId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<{
    account: SocialAccount;
    local_only: true;
    meta_authorization_revoked: false;
  }>(
    `/social-accounts/instagram/${accountId}/disconnect`,
    { product_id: productId, confirm_disconnect: true },
    { signal },
  );
  return response.data.account;
}

export async function connectTikTok(productId: number, signal?: AbortSignal) {
  const response = await apiClient.post<TikTokConnectResult>(
    "/social-accounts/tiktok/connect",
    { product_id: productId },
    { signal },
  );
  return response.data;
}

export async function disconnectTikTokAccount(
  productId: number,
  accountId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<{
    account: SocialAccount;
    local_only: true;
    tiktok_authorization_revoked: false;
  }>(
    `/social-accounts/tiktok/${accountId}/disconnect`,
    { product_id: productId, confirm_disconnect: true },
    { signal },
  );
  return response.data.account;
}

export async function listSocialAccounts(
  productId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<SocialAccount[]>("/social-accounts", {
    params: { product_id: productId },
    signal,
  });
  return response.data;
}

export async function disconnectSocialAccount(
  productId: number,
  accountId: number,
  revokeGoogleAuthorization: boolean,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<{ account: SocialAccount }>(
    `/social-accounts/${accountId}/disconnect`,
    {
      product_id: productId,
      revoke_google_authorization: revokeGoogleAuthorization,
      confirm_disconnect: true,
    },
    { signal },
  );
  return response.data.account;
}

export async function listPublishArtifacts(
  productId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<PublishArtifactCandidate[]>(
    `/products/${productId}/publishing/youtube/artifacts`,
    { signal },
  );
  return response.data;
}

export async function preflightYouTubePublish(
  productId: number,
  data: YouTubePublishingMetadata,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<YouTubePreflight>(
    `/products/${productId}/publishing/youtube/preflight`,
    data,
    { signal },
  );
  return response.data;
}

export async function publishYouTube(
  productId: number,
  data: YouTubePublishingMetadata & {
    preflight_digest: string;
    preflight_expires_at: string;
    input_digest: string;
    idempotency_key: string;
    confirm_upload: true;
  },
  signal?: AbortSignal,
) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/products/${productId}/publishing/youtube`,
    data,
    { signal },
  );
  return response.data;
}

export async function listPublishTasks(
  productId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<PublishTask[]>(
    `/products/${productId}/publish-tasks`,
    { signal },
  );
  return response.data;
}

export async function refreshPublishTask(
  productId: number,
  taskId: number,
  refreshRequestId: string,
  signal?: AbortSignal,
) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/publish-tasks/${taskId}/refresh`,
    { product_id: productId, refresh_request_id: refreshRequestId },
    { signal },
  );
  return response.data;
}

export async function getPublishTask(
  productId: number,
  taskId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<PublishTask>(
    `/publish-tasks/${taskId}`,
    { params: { product_id: productId }, signal },
  );
  return response.data;
}

export async function listYouTubePublishJobs(
  jobType: string,
  sourceType: string,
  sourceId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<ExecutionJob[]>("/execution-jobs", {
    params: { job_type: jobType, source_type: sourceType, source_id: sourceId },
    signal,
  });
  return response.data;
}

export async function getYouTubePublishJob(
  jobId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<ExecutionJob>(
    `/execution-jobs/${jobId}`,
    { signal },
  );
  return response.data;
}

export async function listInstagramPublishArtifacts(productId: number, signal?: AbortSignal) {
  const response = await apiClient.get<PublishArtifactCandidate[]>(
    `/products/${productId}/publishing/instagram/artifacts`, { signal },
  );
  return response.data;
}

export async function preflightInstagramPublish(
  productId: number, data: InstagramPublishingMetadata, signal?: AbortSignal,
) {
  const response = await apiClient.post<InstagramPublishPreflight>(
    `/products/${productId}/publishing/instagram/preflight`, data, { signal },
  );
  return response.data;
}

export async function publishInstagram(
  productId: number,
  data: InstagramPublishingMetadata & {
    input_digest: string; preflight_digest: string; preflight_expires_at: string;
    idempotency_key: string; confirm_upload: true;
  }, signal?: AbortSignal,
) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/products/${productId}/publishing/instagram`, data, { signal },
  );
  return response.data;
}

export async function refreshInstagramPublish(
  productId: number, taskId: number, refreshRequestId: string, signal?: AbortSignal,
) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/publish-tasks/${taskId}/instagram/refresh`,
    { product_id: productId, refresh_request_id: refreshRequestId }, { signal },
  );
  return response.data;
}

export async function preflightInstagramFinalize(
  productId: number, taskId: number, signal?: AbortSignal,
) {
  const response = await apiClient.post<InstagramFinalizePreflight>(
    `/publish-tasks/${taskId}/instagram/finalize-preflight`, null,
    { params: { product_id: productId }, signal },
  );
  return response.data;
}

export async function finalizeInstagramPublish(
  taskId: number,
  data: { product_id: number; finalize_request_id: string; input_digest: string;
    preflight_digest: string; preflight_expires_at: string; confirm_public_publish: true },
  signal?: AbortSignal,
) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/publish-tasks/${taskId}/instagram/finalize`, data, { signal },
  );
  return response.data;
}

export async function listInstagramPublishJobs(
  jobType: string, sourceType: string, sourceId: number, signal?: AbortSignal,
) {
  const response = await apiClient.get<ExecutionJob[]>("/execution-jobs", {
    params: { job_type: jobType, source_type: sourceType, source_id: sourceId }, signal,
  });
  return response.data;
}

export const getInstagramPublishJob = getYouTubePublishJob;

export async function queryTikTokCreatorInfo(productId: number, socialAccountId: number, signal?: AbortSignal) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/products/${productId}/publishing/tiktok/creator-info`,
    { social_account_id: socialAccountId, request_id: crypto.randomUUID() },
    { signal });
  return response.data;
}

export async function getTikTokCreatorInfoSnapshot(
  snapshotId: number, productId: number, socialAccountId: number,
  signal?: AbortSignal,
) {
  const response = await apiClient.get<TikTokCreatorInfoSnapshot>(
    `/tiktok-creator-info-snapshots/${snapshotId}`,
    { params: { product_id: productId, social_account_id: socialAccountId }, signal });
  return response.data;
}

export async function listTikTokPublishArtifacts(productId: number, signal?: AbortSignal) {
  const response = await apiClient.get<PublishArtifactCandidate[]>(
    `/products/${productId}/publishing/tiktok/artifacts`, { signal });
  return response.data;
}

export async function preflightTikTokPublish(productId: number, data: TikTokPublishingMetadata, signal?: AbortSignal) {
  const response = await apiClient.post<TikTokPublishPreflight>(
    `/products/${productId}/publishing/tiktok/preflight`, data, { signal });
  return response.data;
}

export async function publishTikTok(productId: number, data: TikTokPublishingMetadata & {
  input_digest: string; preflight_digest: string; preflight_expires_at: string; confirm_upload: true;
}, signal?: AbortSignal) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/products/${productId}/publishing/tiktok`, data, { signal });
  return response.data;
}

export async function refreshTikTokPublish(productId: number, taskId: number, refreshRequestId: string, signal?: AbortSignal) {
  const response = await apiClient.post<ExecutionJobCreateResult>(
    `/publish-tasks/${taskId}/tiktok/refresh`,
    { product_id: productId, refresh_request_id: refreshRequestId }, { signal });
  return response.data;
}

File: backend/app/api/v1/routes/social.py
import secrets
from typing import Annotated

from fastapi import APIRouter, Depends, Query, Request, Response
from fastapi.responses import RedirectResponse
from sqlalchemy.orm import Session

from app.api.dependencies import (
    BindingInstagramProviderDep,
    BindingTikTokProviderDep,
    BindingYouTubeProviderDep,
    InstagramMediaProbeDep,
    InstagramPublishingGateDep,
    VideoArtifactStorageDep,
    YouTubePublishingGateDep,
)
from app.core.config import Settings, get_settings
from app.db.session import get_db
from app.schemas.execution import ExecutionJobCreateRead
from app.schemas.social import (
    DisconnectRead,
    DisconnectRequest,
    InstagramConnectRead,
    InstagramConnectRequest,
    InstagramDisconnectRead,
    InstagramDisconnectRequest,
    InstagramFinalizePreflightRead,
    InstagramFinalizeRequest,
    InstagramPublishingMetadata,
    InstagramPublishRequest,
    InstagramSubmitPreflightRead,
    PublishArtifactCandidateRead,
    PublishTaskIdentityRequest,
    PublishTaskRead,
    SocialAccountRead,
    TikTokConnectRead,
    TikTokConnectRequest,
    TikTokDisconnectRead,
    TikTokDisconnectRequest,
    YouTubeConnectRead,
    YouTubeConnectRequest,
    YouTubePreflightRead,
    YouTubePublishingMetadata,
    YouTubePublishRequest,
)
from app.services.instagram_account_service import InstagramAccountService
from app.services.instagram_publish_job_service import InstagramPublishJobService
from app.services.instagram_publish_preflight import InstagramPublishPreflightService
from app.services.instagram_publish_service import InstagramPublishService
from app.services.social_security import digest_oauth_state
from app.services.social_service import (
    SocialAccountService,
    YouTubePublishingService,
)
from app.services.tiktok_account_service import TikTokAccountService
from app.services.youtube_publish_job_service import YouTubePublishJobService

router = APIRouter()
OAUTH_BROWSER_COOKIE = "social_oauth_browser"
DbSession = Annotated[Session, Depends(get_db)]
SettingsDep = Annotated[Settings, Depends(get_settings)]


@router.post("/social-accounts/youtube/connect", response_model=YouTubeConnectRead)
def connect_youtube(
    data: YouTubeConnectRequest,
    request: Request,
    response: Response,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingYouTubeProviderDep,
) -> YouTubeConnectRead:
    browser_session = request.cookies.get(OAUTH_BROWSER_COOKIE)
    if not browser_session:
        browser_session = secrets.token_urlsafe(32)
        response.set_cookie(
            OAUTH_BROWSER_COOKIE,
            browser_session,
            max_age=600,
            httponly=True,
            secure=False,
            samesite="lax",
            path="/api/v1/social-accounts",
        )
    return SocialAccountService(db, settings, provider).connect(
        data.product_id,
        browser_session_digest=digest_oauth_state(browser_session),
    )


@router.get("/social-accounts/youtube/callback", response_class=RedirectResponse)
async def youtube_callback(
    request: Request,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingYouTubeProviderDep,
    state: str = Query(min_length=20, max_length=200),
    code: str | None = Query(default=None, max_length=2000),
    error: str | None = Query(default=None, max_length=200),
) -> RedirectResponse:
    location = await SocialAccountService(db, settings, provider).callback(
        state=state,
        browser_session_digest=digest_oauth_state(
            request.cookies.get(OAUTH_BROWSER_COOKIE, "")
        ),
        code=code,
        error=error,
    )
    return RedirectResponse(location, status_code=303)


@router.post("/social-accounts/instagram/connect", response_model=InstagramConnectRead)
def connect_instagram(
    data: InstagramConnectRequest,
    request: Request,
    response: Response,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingInstagramProviderDep,
) -> InstagramConnectRead:
    browser_session = request.cookies.get(OAUTH_BROWSER_COOKIE)
    if not browser_session:
        browser_session = secrets.token_urlsafe(32)
        response.set_cookie(
            OAUTH_BROWSER_COOKIE,
            browser_session,
            max_age=600,
            httponly=True,
            secure=False,
            samesite="lax",
            path="/api/v1/social-accounts",
        )
    return InstagramAccountService(db, settings, provider).connect(
        data.product_id,
        browser_session_digest=digest_oauth_state(browser_session),
    )


@router.get("/social-accounts/instagram/callback", response_class=RedirectResponse)
async def instagram_callback(
    request: Request,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingInstagramProviderDep,
    state: str = Query(min_length=20, max_length=200),
    code: str | None = Query(default=None, max_length=2000),
    error: str | None = Query(default=None, max_length=200),
) -> RedirectResponse:
    location = await InstagramAccountService(db, settings, provider).callback(
        state=state,
        browser_session_digest=digest_oauth_state(
            request.cookies.get(OAUTH_BROWSER_COOKIE, "")
        ),
        code=code,
        error=error,
    )
    return RedirectResponse(location, status_code=303)


@router.post(
    "/social-accounts/instagram/{account_id}/disconnect",
    response_model=InstagramDisconnectRead,
)
def disconnect_instagram_account(
    account_id: int,
    data: InstagramDisconnectRequest,
    db: DbSession,
    settings: SettingsDep,
) -> InstagramDisconnectRead:
    return InstagramAccountService(db, settings, None).disconnect(
        account_id, product_id=data.product_id
    )


@router.post("/social-accounts/tiktok/connect", response_model=TikTokConnectRead)
def connect_tiktok(
    data: TikTokConnectRequest,
    request: Request,
    response: Response,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingTikTokProviderDep,
) -> TikTokConnectRead:
    browser_session = request.cookies.get(OAUTH_BROWSER_COOKIE)
    if not browser_session:
        browser_session = secrets.token_urlsafe(32)
        response.set_cookie(
            OAUTH_BROWSER_COOKIE,
            browser_session,
            max_age=600,
            httponly=True,
            secure=False,
            samesite="lax",
            path="/api/v1/social-accounts",
        )
    return TikTokAccountService(db, settings, provider).connect(
        data.product_id,
        browser_session_digest=digest_oauth_state(browser_session),
    )


@router.get("/social-accounts/tiktok/callback", response_class=RedirectResponse)
async def tiktok_callback(
    request: Request,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingTikTokProviderDep,
    state: str = Query(min_length=20, max_length=200),
    code: str | None = Query(default=None, max_length=2000),
    error: str | None = Query(default=None, max_length=200),
) -> RedirectResponse:
    location = await TikTokAccountService(db, settings, provider).callback(
        state=state,
        browser_session_digest=digest_oauth_state(
            request.cookies.get(OAUTH_BROWSER_COOKIE, "")
        ),
        code=code,
        error=error,
    )
    return RedirectResponse(location, status_code=303)


@router.post(
    "/social-accounts/tiktok/{account_id}/disconnect",
    response_model=TikTokDisconnectRead,
)
def disconnect_tiktok_account(
    account_id: int,
    data: TikTokDisconnectRequest,
    db: DbSession,
    settings: SettingsDep,
) -> TikTokDisconnectRead:
    return TikTokAccountService(db, settings, None).disconnect(
        account_id, product_id=data.product_id
    )


@router.get("/social-accounts", response_model=list[SocialAccountRead])
def list_social_accounts(
    db: DbSession,
    settings: SettingsDep,
    product_id: int = Query(gt=0),
) -> list[SocialAccountRead]:
    return SocialAccountService(db, settings, None).list_accounts(product_id)


@router.get("/social-accounts/{account_id}", response_model=SocialAccountRead)
def get_social_account(
    account_id: int,
    db: DbSession,
    settings: SettingsDep,
    product_id: int = Query(gt=0),
) -> SocialAccountRead:
    return SocialAccountService(db, settings, None).get_account(account_id, product_id)


@router.post("/social-accounts/{account_id}/disconnect", response_model=DisconnectRead)
async def disconnect_social_account(
    account_id: int,
    data: DisconnectRequest,
    db: DbSession,
    settings: SettingsDep,
    provider: BindingYouTubeProviderDep,
) -> DisconnectRead:
    return await SocialAccountService(db, settings, provider).disconnect(
        account_id,
        product_id=data.product_id,
        revoke_google_authorization=data.revoke_google_authorization,
    )


@router.get(
    "/products/{product_id}/publishing/youtube/artifacts",
    response_model=list[PublishArtifactCandidateRead],
)
def list_youtube_publish_artifacts(
    product_id: int,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
) -> list[PublishArtifactCandidateRead]:
    return YouTubePublishingService(db, settings, None, storage).list_candidates(
        product_id
    )


@router.post(
    "/products/{product_id}/publishing/youtube/preflight",
    response_model=YouTubePreflightRead,
)
def preflight_youtube_publish(
    product_id: int,
    data: YouTubePublishingMetadata,
    db: DbSession,
    settings: SettingsDep,
    gate: YouTubePublishingGateDep,
    storage: VideoArtifactStorageDep,
) -> YouTubePreflightRead:
    del gate
    return YouTubePublishingService(db, settings, None, storage).preflight(
        product_id, data
    )


@router.post(
    "/products/{product_id}/publishing/youtube",
    response_model=ExecutionJobCreateRead,
    status_code=201,
)
def publish_youtube_video(
    product_id: int,
    data: YouTubePublishRequest,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
) -> ExecutionJobCreateRead:
    return YouTubePublishJobService(db, settings, storage).enqueue_submit(
        product_id, data
    )


@router.get("/publish-tasks/{task_id}", response_model=PublishTaskRead)
def get_publish_task(
    task_id: int,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    product_id: int = Query(gt=0),
) -> PublishTaskRead:
    return YouTubePublishingService(db, settings, None, storage).get_task(
        task_id, product_id
    )


@router.get(
    "/products/{product_id}/publish-tasks", response_model=list[PublishTaskRead]
)
def list_publish_tasks(
    product_id: int,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
) -> list[PublishTaskRead]:
    return YouTubePublishingService(db, settings, None, storage).list_tasks(product_id)


@router.post(
    "/publish-tasks/{task_id}/refresh",
    response_model=ExecutionJobCreateRead,
    status_code=201,
)
def refresh_publish_task(
    task_id: int,
    data: PublishTaskIdentityRequest,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
) -> ExecutionJobCreateRead:
    return YouTubePublishJobService(db, settings, storage).enqueue_refresh(
        task_id, data
    )


@router.get(
    "/products/{product_id}/publishing/instagram/artifacts",
    response_model=list[PublishArtifactCandidateRead],
)
def list_instagram_publish_artifacts(
    product_id: int,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> list[PublishArtifactCandidateRead]:
    del gate
    return InstagramPublishPreflightService(
        db, settings, storage, probe
    ).list_candidates(product_id)


@router.post(
    "/products/{product_id}/publishing/instagram/preflight",
    response_model=InstagramSubmitPreflightRead,
)
def preflight_instagram_publish(
    product_id: int,
    data: InstagramPublishingMetadata,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> InstagramSubmitPreflightRead:
    del gate
    return InstagramPublishPreflightService(db, settings, storage, probe).run(
        product_id, data
    )


@router.post(
    "/products/{product_id}/publishing/instagram",
    response_model=ExecutionJobCreateRead,
    status_code=201,
)
def publish_instagram_reel(
    product_id: int,
    data: InstagramPublishRequest,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> ExecutionJobCreateRead:
    del gate
    return InstagramPublishJobService(db, settings, storage, probe).enqueue_submit(
        product_id, data
    )


@router.post(
    "/publish-tasks/{task_id}/instagram/refresh",
    response_model=ExecutionJobCreateRead,
    status_code=201,
)
def refresh_instagram_publish(
    task_id: int,
    data: PublishTaskIdentityRequest,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> ExecutionJobCreateRead:
    del gate
    return InstagramPublishJobService(db, settings, storage, probe).enqueue_refresh(
        task_id, data
    )


@router.post(
    "/publish-tasks/{task_id}/instagram/finalize-preflight",
    response_model=InstagramFinalizePreflightRead,
)
def preflight_instagram_finalize(
    task_id: int,
    product_id: int,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> InstagramFinalizePreflightRead:
    del gate
    return InstagramPublishService(db, settings, storage, probe).finalize_preflight(
        task_id, product_id
    )


@router.post(
    "/publish-tasks/{task_id}/instagram/finalize",
    response_model=ExecutionJobCreateRead,
    status_code=201,
)
def finalize_instagram_publish(
    task_id: int,
    data: InstagramFinalizeRequest,
    db: DbSession,
    settings: SettingsDep,
    storage: VideoArtifactStorageDep,
    probe: InstagramMediaProbeDep,
    gate: InstagramPublishingGateDep,
) -> ExecutionJobCreateRead:
    del gate
    return InstagramPublishJobService(db, settings, storage, probe).enqueue_finalize(
        task_id, data
    )
